compose-settings-storage-datastore/src/main/java/com/alorma/compose/settings/storage/datastore/DataStoreAsyncImpl.kt

```diff
@@ -0,0 +1,122 @@
+                SettingsSnapshotLoadState.Success(DataStoreSnapshot(it))
+            }.onStart {
+                emit(SettingsSnapshotLoadState.Loading())
+            }.catch { cause ->
+                emit(SettingsSnapshotLoadState.Error(cause))
+            }
+        }
+
+    override fun retryLoadSnapshot() {
+        retryTriggerFlow.value = Any()
+    }
+
+    override suspend fun updateSnapshot(block: (MutableDataStoreSnapshot) -> Unit) {
+        dataStore.edit {
+            block(MutableDataStoreSnapshot(it))
+        }
+    }
+}
+
+fun<T> Preferences.Key<T>.asSettingsSnapshotKey(defaultValue: T): DataStoreSnapshot.Key<T> {
+    return DataStoreSnapshot.Key(this, defaultValue)
+}
+
+fun dataStoreIntSettingsSnapshotKey(name: String, defaultValue: Int): DataStoreSnapshot.Key<Int> {
+    return intPreferencesKey(name).asSettingsSnapshotKey(defaultValue)
+}
+
+fun dataStoreLongSettingsSnapshotKey(name: String, defaultValue: Long): DataStoreSnapshot.Key<Long> {
+    return longPreferencesKey(name).asSettingsSnapshotKey(defaultValue)
+}
+
+fun dataStoreFloatSettingsSnapshotKey(name: String, defaultValue: Float): DataStoreSnapshot.Key<Float> {
+    return floatPreferencesKey(name).asSettingsSnapshotKey(defaultValue)
+}
+
+fun dataStoreDoubleSettingsSnapshotKey(name: String, defaultValue: Double): DataStoreSnapshot.Key<Double> {
+    return doublePreferencesKey(name).asSettingsSnapshotKey(defaultValue)
+}
+
+fun dataStoreBooleanSettingsSnapshotKey(name: String, defaultValue: Boolean): DataStoreSnapshot.Key<Boolean> {
+    return booleanPreferencesKey(name).asSettingsSnapshotKey(defaultValue)
+}
+
+fun dataStoreStringSettingsSnapshotKey(name: String, defaultValue: String): DataStoreSnapshot.Key<String> {
+    return stringPreferencesKey(name).asSettingsSnapshotKey(defaultValue)
+}
+
+fun dataStoreStringSetSettingsSnapshotKey(name: String, defaultValue: Set<String>): DataStoreSnapshot.Key<Set<String>> {
+    return stringSetPreferencesKey(name).asSettingsSnapshotKey(defaultValue)
+}
+
+class DataStoreSnapshotKeySelector(private val name: String) {
+    class Result<T> internal constructor (val preferencesKey: Preferences.Key<T>, val defaultValue: T)
+
+    fun int(defaultValue: Int) = result(defaultValue)
+    fun long(defaultValue: Long) = result(defaultValue)
+    fun boolean(defaultValue: Boolean) = result(defaultValue)
+    fun float(defaultValue: Float) = result(defaultValue)
+    fun double(defaultValue: Double) = result(defaultValue)
+    fun string(defaultValue: String) = result(defaultValue)
+    fun stringSet(defaultValue: Set<String>) = result(defaultValue)
+
+    @Suppress("UNCHECKED_CAST")
+    private fun<T> result(defaultValue: T): Result<T> {
+        return Result(intPreferencesKey(name) as Preferences.Key<T>, defaultValue)
+    }
+}
+
+inline fun<T> dataStoreSnapshotKey(name: String, resolver: DataStoreSnapshotKeySelector.() -> DataStoreSnapshotKeySelector.Result<T>): DataStoreSnapshot.Key<T> {
```

Comment:
I haven't decided which version of key creation is better to use (one with `dataStore*type*settingsSnapshotKey` or `dataStoreSnapshotKey` builder). @alorma What do you think?